### PR TITLE
Смирнов -- Лабораторная работа #2

### PR DESCRIPTION
**Pull request by AndrewSmrnv**
@VladimirTyagunov , @KalyanovDO , could you please review lab 2?

### CODE REVIEW

File: code/string-calculator/viewmodel/src/main/java/ru/unn/agile/calc/viewmodel/StringCalcViewModel.java [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+                String calcResult = String.valueOf(calculator.result(expressionTf.get()));
+                resultLbl.setValue((calcResult));
+            } catch (Exception e) {
+                resultLbl.setValue(ERROR_MESSAGE);
+            }
+        }
+    }
+
+    public boolean isExpressionValid() {
+        String exprText = expressionTf.get().trim();
```
Comment (vtyagunov) [inline code/string-calculator/viewmodel/src/main/java/ru/unn/agile/calc/viewmodel/StringCalcViewModel.java]: use replaceAll for removing all spaces.

File: code/string-calculator/viewmodel/src/main/java/ru/unn/agile/calc/viewmodel/StringCalcViewModel.java [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+            }
+        };
+        calculationDisabled.bind(canCalculateBoolBinding.not());
+    }
+
+    public void calculate() {
+        if (!isCalculationDisabled()) {
+            try {
+                String calcResult = String.valueOf(calculator.result(expressionTf.get()));
+                resultLbl.setValue((calcResult));
```
Comment (vtyagunov) [inline code/string-calculator/viewmodel/src/main/java/ru/unn/agile/calc/viewmodel/StringCalcViewModel.java]: use replaceAll for removing all spaces.

File: code/string-calculator/view/src/main/java/ru/unn/agile/calc/view/StringCalcView.java [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+import javafx.scene.control.Button;
+import javafx.scene.control.Label;
+import javafx.scene.control.TextField;
+import javafx.scene.control.Tooltip;
+import ru.unn.agile.calc.viewmodel.StringCalcViewModel;
+
+public class StringCalcView {
+
+    private final PseudoClass errorClass = PseudoClass.getPseudoClass("error");
+    private final String tooltip = "Type your expression here";
```
Comment (KalyanovDO) [inline code/string-calculator/view/src/main/java/ru/unn/agile/calc/view/StringCalcView.java]: Please change to the clear description of valid input

**Review by vtyagunov — CHANGES_REQUESTED** [review]

File: code/string-calculator/view/src/main/java/ru/unn/agile/calc/view/StringCalcView.java (hunk 3 above)
Comment (AndrewSmrnv) [reply code/string-calculator/view/src/main/java/ru/unn/agile/calc/view/StringCalcView.java]: fixed

File: code/string-calculator/viewmodel/src/main/java/ru/unn/agile/calc/viewmodel/StringCalcViewModel.java (hunk 2 above)
Comment (AndrewSmrnv) [reply code/string-calculator/viewmodel/src/main/java/ru/unn/agile/calc/viewmodel/StringCalcViewModel.java]: fixed

File: code/string-calculator/viewmodel/src/main/java/ru/unn/agile/calc/viewmodel/StringCalcViewModel.java (hunk 1 above)
Comment (AndrewSmrnv) [reply code/string-calculator/viewmodel/src/main/java/ru/unn/agile/calc/viewmodel/StringCalcViewModel.java]: fixed

**Comment by vtyagunov** [comment]
👍

**Comment by KalyanovDO** [comment]
:+1:

**Pull request merged**